_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+          <ul>
+            <li>Item 1</li>
+            <li>Item 2</li>
+            <li>Item 3</li>
+          </ul>
+        </Drawer>
+      </div>
+    );
+  }
+}
+```
+
+Here's what it looks like so far. Really the only thing it does is provide
+structure and toggle a CSS class `is-expanded` based on state.
+
+![Unstyled Drawer Component](/img/2019/unstyled-drawer-component.gif)
+
+It's pretty ugly at this point, so in the next section we're going to add some
+SCSS styles to it.
+
+## Adding Style
+
+One thing missing from our component at the moment is that the Drawer's content
+is always shown. We only want it to display if the state of the component is expanded. We're going to eventually animate it into existing using a slide transition so focusing on the element's height will be advantageous here.
+
+Start off by hiding the content using the following style
+configuration.
+
+``` scss
+.Drawer {
+  &-content {
+    height: 0;
+    overflow: hidden;
+  }
+}
+```
+
+We need both `height: 0` and `overflow: hidden` here as they work in tandem. Having a
+height of 0 still shows the content even with a container without height. Adding overflow hidden ensures that the content in the container outside of the height will remain hidden. These two together remove the content.
+
+Next we probably want the `Drawer-title` element to look like it is actually clickable. We can do this by adding a border, background, and cursor styles below.
+
+``` scss
+.Drawer {
+  &-title {
+    background: #efefef;
+    border: 1px solid;
+    cursor: pointer;
+    display: inline-block;
+    padding: 16px;
+  }
+
+  &-content {
+    ...
+  }
+}
+```
+
+The result is a primitive looking button to click. Which brings up something else we should probably fix at this point which is keyboard accessibility. Time for a slight detour.
+
+![Basic title styles](/img/2019/basic-title-style.png)
+
+### Keyboard Accessibility
+
+Right now on the page anyone using a screen reader or other non-mouse assistance technology, will be unable to tab to our new button. One way we could attempt to fix this is by giving our `Drawer-title` div a `tabIndex` value and a `role` value.
+
+``` jsx
+<div
+  className="Drawer-title"
+  onClick={this.handleTitleClick}
+  tabIndex="0"
+  role="button"
+>
+  {title}
+</div>
+```
+
+<blockquote class="Info Info-right">
+  The eslint plugin jsx-a11y has some great documentation on the subject of non-semantic
+  elements having an onClick handler.
+  <cite><a href="https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md" target="_blank">ESlint jsx-a11y</a></cite>
+</blockquote>
+
+While, this allows us to focus on the element, the enter and spacebar keys don't activate our `onClick` handler. Major bummer. We could try to add an `onKeyPress` event and capture the spacebar and enter key, but why do more than we need to when there is a perfectly semantic approach. The `<button>` element.
+
```

Comment:
explain what this means in terms of access. ie "allows us to focus, that means the screen reader could read the word. but the enter and space bar keys don't active it so basically our user could see the drawer but not open it.